File: binsV1/binV3/first_problem.rb
require 'json'

input_data = [
  { "name": "Alice", "email": "alice@gmail.com" },
  { "name": "Bob", "email": "bob@yahoo.com" },
  { "name": "Carol", "email": "carol@gmail.com" }
]

grouped_domains = input_data.group_by { |data| data[:email].split('@')[1] }

filtered_names = grouped_domains.each_with_object({}) { |(k, v), h| h[k] = v.map { |a| a[:name] } }

print JSON.pretty_generate(filtered_names)

File: bins/binsV1/binV9/first_problem.js
const users = [
  { name: "Alice", email: "alice@gmail.com" },
  { name: "Bob", email: "bob@yahoo.com" },
  { name: "Carol", email: "carol@gmail.com" },
];

const groupedUsers = Object.groupBy(users, (user) => {
  return user.email.split("@")[1];
});

const filteredUsers = Object.fromEntries(
  Object.entries(groupedUsers).map(([domain, users]) => {
    return [domain, users.map((user) => user.name)];
  })
);

console.log(JSON.stringify({ data: filteredUsers }, null, 2));
